File: typescript/src/resources/webhooks.ts
/**
 * inkbox-mail/resources/webhooks.ts
 *
 * Webhook CRUD for mailboxes.
 */

import { HttpTransport } from "../_http.js";
import {
  Webhook,
  WebhookCreateResult,
  RawWebhook,
  RawWebhookCreateResult,
  parseWebhook,
  parseWebhookCreateResult,
} from "../types.js";

export class WebhooksResource {
  constructor(private readonly http: HttpTransport) {}

  /**
   * Create a webhook subscription for a mailbox.
   *
   * @param mailboxId - Email address or UUID of the mailbox.
   * @param options.url - HTTPS URL to receive webhook events.
   * @param options.eventTypes - List of event types to subscribe to.
   */
  async create(
    mailboxId: string,
    options: { url: string; eventTypes: string[] },
  ): Promise<WebhookCreateResult> {
    const data = await this.http.post<RawWebhookCreateResult>(
      `/mailboxes/${mailboxId}/webhooks`,
      { url: options.url, event_types: options.eventTypes },
    );
    return parseWebhookCreateResult(data);
  }

  /** List all webhooks for a mailbox. */
  async list(mailboxId: string): Promise<Webhook[]> {
    const data = await this.http.get<RawWebhook[]>(
      `/mailboxes/${mailboxId}/webhooks`,
    );
    return data.map(parseWebhook);
  }

  /** Delete a webhook subscription. */
  async delete(mailboxId: string, webhookId: string): Promise<void> {
    await this.http.delete(`/mailboxes/${mailboxId}/webhooks/${webhookId}`);
  }
}

File: python/inkbox/mail/resources/webhooks.py
"""
inkbox/mail/resources/webhooks.py

Webhook CRUD for mailboxes.
"""

from __future__ import annotations

from typing import TYPE_CHECKING

from inkbox.mail.types import Webhook, WebhookCreateResult

if TYPE_CHECKING:
    from inkbox.mail._http import HttpTransport


class WebhooksResource:
    def __init__(self, http: HttpTransport) -> None:
        self._http = http

    def create(
        self,
        mailbox_id: str,
        *,
        url: str,
        event_types: list[str],
    ) -> WebhookCreateResult:
        """Create a webhook subscription for a mailbox.

        Args:
            mailbox_id: UUID of the mailbox.
            url: HTTPS URL to receive webhook events.
            event_types: List of event types to subscribe to.

        Returns:
            The created webhook, including the signing secret.
        """
        data = self._http.post(
            f"/mailboxes/{mailbox_id}/webhooks",
            json={"url": url, "event_types": event_types},
        )
        return WebhookCreateResult._from_dict(data)

    def list(self, mailbox_id: str) -> list[Webhook]:
        """List all webhooks for a mailbox.

        Args:
            mailbox_id: UUID of the mailbox.
        """
        data = self._http.get(f"/mailboxes/{mailbox_id}/webhooks")
        return [Webhook._from_dict(w) for w in data]

    def delete(self, mailbox_id: str, webhook_id: str) -> None:
        """Delete a webhook subscription.

        Args:
            mailbox_id: UUID of the mailbox.
            webhook_id: UUID of the webhook to delete.
        """
        self._http.delete(f"/mailboxes/{mailbox_id}/webhooks/{webhook_id}")
